File: web/src/repo/linked-tasks.ts
import { api } from './client'

// ─── Enums ──────────────────────────────────────────────────────────────────

export type LinkedTaskStatus = 'IDLE' | 'RUNNING' | 'SUCCESS' | 'FAILED'
export type LinkedTaskRunStatus = 'RUNNING' | 'SUCCESS' | 'FAILED' | 'CANCELLED'
export type LinkedTaskStepStatus = 'PENDING' | 'RUNNING' | 'SUCCESS' | 'FAILED' | 'SKIPPED'
export type EdgeCondition = 'ON_SUCCESS' | 'ALWAYS'

// ─── Data types ──────────────────────────────────────────────────────────────

export interface FlowTaskRef {
    id: number
    name: string
    status: string
}

export interface LinkedTaskStep {
    id: number
    linked_task_id: number
    flow_task_id: number
    pos_x: number
    pos_y: number
    flow_task?: FlowTaskRef
    created_at: string
    updated_at: string
}

export interface LinkedTaskEdge {
    id: number
    linked_task_id: number
    source_step_id: number
    target_step_id: number
    condition: EdgeCondition
    created_at: string
    updated_at: string
}

export interface LinkedTask {
    id: number
    name: string
    description: string | null
    status: LinkedTaskStatus
    last_run_at: string | null
    last_run_status: LinkedTaskRunStatus | null
    created_at: string
    updated_at: string
}

export interface LinkedTaskDetail extends LinkedTask {
    steps: LinkedTaskStep[]
    edges: LinkedTaskEdge[]
}

export interface LinkedTaskListResponse {
    items: LinkedTask[]
    total: number
    page: number
    page_size: number
}

// ─── Graph save ──────────────────────────────────────────────────────────────

export interface LinkedTaskStepSave {
    id?: number | string
    flow_task_id: number
    pos_x: number
    pos_y: number
}

export interface LinkedTaskEdgeSave {
    source_step_id: number | string
    target_step_id: number | string
    condition: EdgeCondition
}

export interface LinkedTaskGraphSave {
    steps: LinkedTaskStepSave[]
    edges: LinkedTaskEdgeSave[]
}

// ─── Run history ─────────────────────────────────────────────────────────────

export interface LinkedTaskRunStepLog {
    id: number
    run_history_id: number
    step_id: number
    flow_task_id: number
    flow_task_run_history_id: number | null
    status: LinkedTaskStepStatus
    started_at: string | null
    finished_at: string | null
    error_message: string | null
    created_at: string
}

export interface LinkedTaskRunHistory {
    id: number
    linked_task_id: number
    trigger_type: string
    status: LinkedTaskRunStatus
    celery_task_id: string | null
    started_at: string
    finished_at: string | null
    error_message: string | null
    step_logs: LinkedTaskRunStepLog[]
    created_at: string
}

export interface LinkedTaskRunHistoryListResponse {
    items: LinkedTaskRunHistory[]
    total: number
    page: number
    page_size: number
}

// ─── Trigger ─────────────────────────────────────────────────────────────────

export interface LinkedTaskTriggerResponse {
    message: string
    run_id: number
    celery_task_id: string
    status: string
}

// ─── Repository ──────────────────────────────────────────────────────────────

export const linkedTasksRepo = {
    list(page = 1, pageSize = 20) {
        return api.get<LinkedTaskListResponse>('/linked-tasks', {
            params: { page, page_size: pageSize },
        })
    },

    get(id: number) {
        return api.get<LinkedTaskDetail>(`/linked-tasks/${id}`)
    },

    create(payload: { name: string; description?: string }) {
        return api.post<LinkedTask>('/linked-tasks', payload)
    },

    update(id: number, payload: { name?: string; description?: string }) {
        return api.put<LinkedTask>(`/linked-tasks/${id}`, payload)
    },

    remove(id: number) {
        return api.delete<void>(`/linked-tasks/${id}`)
    },

    saveGraph(id: number, graph: LinkedTaskGraphSave) {
        return api.post<LinkedTaskDetail>(`/linked-tasks/${id}/graph`, graph)
    },

    trigger(id: number) {
        return api.post<LinkedTaskTriggerResponse>(`/linked-tasks/${id}/run`)
    },

    getRuns(id: number, page = 1, pageSize = 20) {
        return api.get<LinkedTaskRunHistoryListResponse>(`/linked-tasks/${id}/runs`, {
            params: { page, page_size: pageSize },
        })
    },
}

File: backend/app/api/v1/endpoints/linked_tasks.py
"""
Linked Task API endpoints.

REST API for managing linked task orchestration DAGs,
graph save/load, run triggering, and history.
"""

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_linked_task_service
from app.core.exceptions import EntityNotFoundError
from app.core.logging import get_logger
from app.domain.schemas.linked_task import (
    LinkedTaskCreate,
    LinkedTaskDetailResponse,
    LinkedTaskEdgeResponse,
    LinkedTaskGraphSave,
    LinkedTaskListResponse,
    LinkedTaskResponse,
    LinkedTaskRunHistoryListResponse,
    LinkedTaskRunHistoryResponse,
    LinkedTaskStepResponse,
    LinkedTaskTriggerResponse,
    LinkedTaskUpdate,
)
from app.domain.services.linked_task import LinkedTaskService

logger = get_logger(__name__)
router = APIRouter()


# ─── CRUD ─────────────────────────────────────────────────────────────────────

@router.get(
    "",
    response_model=LinkedTaskListResponse,
    summary="List linked tasks",
)
def list_linked_tasks(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskListResponse:
    items, total = service.list_linked_tasks(page, page_size)
    return LinkedTaskListResponse(
        items=[LinkedTaskResponse.from_orm(t) for t in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post(
    "",
    response_model=LinkedTaskResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create linked task",
)
def create_linked_task(
    data: LinkedTaskCreate,
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskResponse:
    try:
        task = service.create_linked_task(data)
        return LinkedTaskResponse.from_orm(task)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get(
    "/{linked_task_id}",
    response_model=LinkedTaskDetailResponse,
    summary="Get linked task",
)
def get_linked_task(
    linked_task_id: int,
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskDetailResponse:
    try:
        task = service.get_linked_task(linked_task_id)
        steps, edges = service.get_graph(linked_task_id)
        resp = LinkedTaskDetailResponse.from_orm(task)
        resp.steps = [LinkedTaskStepResponse.from_orm(s) for s in steps]
        resp.edges = [LinkedTaskEdgeResponse.from_orm(e) for e in edges]
        return resp
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.put(
    "/{linked_task_id}",
    response_model=LinkedTaskResponse,
    summary="Update linked task",
)
def update_linked_task(
    linked_task_id: int,
    data: LinkedTaskUpdate,
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskResponse:
    try:
        task = service.update_linked_task(linked_task_id, data)
        return LinkedTaskResponse.from_orm(task)
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.delete(
    "/{linked_task_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete linked task",
)
def delete_linked_task(
    linked_task_id: int,
    service: LinkedTaskService = Depends(get_linked_task_service),
):
    try:
        service.delete_linked_task(linked_task_id)
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


# ─── Graph ─────────────────────────────────────────────────────────────────────

@router.post(
    "/{linked_task_id}/graph",
    response_model=LinkedTaskDetailResponse,
    summary="Save linked task graph",
)
def save_graph(
    linked_task_id: int,
    data: LinkedTaskGraphSave,
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskDetailResponse:
    """Replace the full step/edge graph for a linked task."""
    try:
        steps, edges = service.save_graph(linked_task_id, data)
        task = service.get_linked_task(linked_task_id)
        resp = LinkedTaskDetailResponse.from_orm(task)
        resp.steps = [LinkedTaskStepResponse.from_orm(s) for s in steps]
        resp.edges = [LinkedTaskEdgeResponse.from_orm(e) for e in edges]
        return resp
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to save linked task graph: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to save graph",
        )


# ─── Run ───────────────────────────────────────────────────────────────────────

@router.post(
    "/{linked_task_id}/run",
    response_model=LinkedTaskTriggerResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Trigger linked task run",
)
def trigger_run(
    linked_task_id: int,
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskTriggerResponse:
    try:
        run = service.trigger_run(linked_task_id)
        return LinkedTaskTriggerResponse(
            message="Linked task dispatched",
            run_id=run.id,
            celery_task_id=run.celery_task_id or "",
            status="RUNNING",
        )
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to trigger linked task {linked_task_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to trigger run",
        )


# ─── Run History ───────────────────────────────────────────────────────────────

@router.get(
    "/{linked_task_id}/runs",
    response_model=LinkedTaskRunHistoryListResponse,
    summary="Get linked task run history",
)
def get_run_history(
    linked_task_id: int,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskRunHistoryListResponse:
    try:
        items, total = service.get_run_history(linked_task_id, page, page_size)
        return LinkedTaskRunHistoryListResponse(
            items=[LinkedTaskRunHistoryResponse.from_orm(r) for r in items],
            total=total,
            page=page,
            page_size=page_size,
        )
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


# ─── Cancel Run ────────────────────────────────────────────────────────────────

@router.post(
    "/{linked_task_id}/runs/{run_id}/cancel",
    response_model=LinkedTaskRunHistoryResponse,
    summary="Cancel a running linked task run",
)
def cancel_run(
    linked_task_id: int,
    run_id: int,
    service: LinkedTaskService = Depends(get_linked_task_service),
) -> LinkedTaskRunHistoryResponse:
    try:
        run = service.cancel_run(linked_task_id, run_id)
        return LinkedTaskRunHistoryResponse.from_orm(run)
    except EntityNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to cancel run {run_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to cancel run",
        )
